rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+
+    somethingElse
+  ];
+
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  inherit
+    lib
+    foo
+    bar
+    baz
+    ;
+}
+```
+
+### Function application
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- As many arguments as possible must be fit onto the first line.
+  - If all but the last argument do fit, then the last argument may also start on the first line.
+  - If an earlier argument does not fit onto the first line, then that argument and all the following ones must start on their own line.
+  - All arguments that are not on the same line as the function must be indented by one level.
+
+**Examples:**
+
+```nix
+# All arguments fit onto the first line
+function arg1 arg2
+
+# The line length limit is reached, so the remaining arguments need to be on their own lines
+function arg1 arg2 arg3
+  arg4
+  arg5
+
+# The last argument is a multiline expression, so it doesn't fit on the first line,
+# but it can still start on the first line
+function arg1 arg2 {
+  more = "things";
+}
+
+# The second argument doesn't fit on the first line, but it's not the last argument,
+# so it needs to start on a new line
+function arg1 arg2
+  {
+    more = "things";
+  }
+  arg3
+
+# Same with more multiline arguments
+function
+  {
+    a = 1;
+    b = 2;
+  }
+  {
+    c = 1;
+    d = 2;
+  }
+```
+
+**Drawbacks**
+
+- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.
+
+**Alternatives**
+
+- Compacting multiline arguments like this:
+  ```nix
+  function arg1 {
+    # stuff
+  } arg3
+
+  function {
+    # …
+  } {
+    # …
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
+
+### Function declaration
+
+- The body of the function must not be indented relative to its first arguments.
+- Multiple ("simple") identifier arguments must be written onto the same line if admitted from on the line length limit.
+  - Otherwise they're each on their own line.
+- Attribute set arguments must always start on a new line and they must not be mixed with identifier arguments.
+  - If they have few attributes, the argument may be written on a single line
+  - Otherwise each attribute must be on its own line with indentation, followed by a trailing comma.
+
+**Examples**
+
+```nix
+#1
+name: value: name ++ value
+
+#2
+name: value:
+{
+  "${name}-foo" = value;
+}
+
+#3
+{ pkgs }: pkgs.hello
+
+#4
+args@{
+  some,
+  argument,
+  default ? value,
+  ...
+}:
+{
+  # body
+}
+
+#5
+{ pkgs }:
+name: value: 
+{
+  # body
+}
+
+#6: These would be over the line length limit on a single line
+aaaa:
+bbbb:
+cccc: 
+dddd:
+eeee:
+null
+
+#7: @ patterns can also come afterwards
+{ pkgs }@args: pkgs.hello
+```
+
+**Alternatives**
+
+- Have leading commas for parameters in attribute set arguments, like currently done in Nixpkgs.
+
```

Comment:
Little sad to see this go, but I think your points are valid in terms of normalization expression expansion.

Great work, having normalized formatting will remove a LOT of bike-shedding.